File: Sample-Code-Snippets/Python/azure-service-bus-emulator-console-sample/sample-send-receive.py
# Install the Azure Service Bus SDK for Python using following command prior to running the sample. Versions >=7.14.0 support Emulator.
# pip install azure-servicebus==7.14.3 

from uuid import uuid4
from azure.servicebus import ServiceBusClient, ServiceBusMessage
from azure.servicebus.management import ServiceBusAdministrationClient
import os
import time
import urllib.request
import urllib.error

EMULATOR_HOST = "localhost"
EMULATOR_HTTP_PORT = 5300
CONNECTION_STR = f"Endpoint=sb://{EMULATOR_HOST};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;"
ADMIN_API_CONNECTION_STR = f"Endpoint=sb://{EMULATOR_HOST}:{EMULATOR_HTTP_PORT};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;"

def wait_for_servicebus_emulator_to_be_healthy():
    health_url = f"http://{EMULATOR_HOST}:{EMULATOR_HTTP_PORT}/health"
    print("Waiting for Service Bus Emulator to be healthy...", flush=True)
    while True:
        try:
            with urllib.request.urlopen(health_url, timeout=3) as resp:
                if resp.status == 200:
                    print("Service Bus Emulator is healthy.", flush=True)
                    return
        except Exception as ex:
            print(f"Service Bus Emulator is not healthy yet. health_url={health_url} Continuing to wait...{ex}", flush=True)
        time.sleep(1)

def update_connection_strings_from_env():
    global CONNECTION_STR, ADMIN_API_CONNECTION_STR
    CONNECTION_STR = f"Endpoint=sb://{EMULATOR_HOST};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;"
    ADMIN_API_CONNECTION_STR = f"Endpoint=sb://{EMULATOR_HOST}:{EMULATOR_HTTP_PORT};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;"
    if EMULATOR_HTTP_PORT == 80:
        ADMIN_API_CONNECTION_STR = CONNECTION_STR
    print(f"Runtime ConnectionString: {CONNECTION_STR}")
    print(f"Admin API ConnectionString: {ADMIN_API_CONNECTION_STR}")

def send_message(sender):
    message = ServiceBusMessage("Hello, Azure Service Bus!")
    sender.send_messages(message)
    print("Message sent:", message)

def receive_messages(receiver):
    with receiver:
        for msg in receiver.receive_messages(max_message_count=1, max_wait_time=5):
            print("Received message:", str(msg))
            receiver.complete_message(msg)

def main():
    global EMULATOR_HOST, EMULATOR_HTTP_PORT
    EMULATOR_HOST = os.getenv("EMULATOR_HOST", "localhost")
    EMULATOR_HTTP_PORT = int(os.getenv("EMULATOR_HTTP_PORT", "5300"))
    
    wait_for_servicebus_emulator_to_be_healthy()
    update_connection_strings_from_env()
    with ServiceBusAdministrationClient.from_connection_string(conn_str=ADMIN_API_CONNECTION_STR) as admin_client:
        # Temporary hack until SDK supports non-TLS and custom-port for management operations
        # Override the base URL to point to the emulator's admin API endpoint
        admin_client._impl._client._base_url = f"http://{EMULATOR_HOST}:{EMULATOR_HTTP_PORT}"

        queue_name = f"queue-{uuid4().hex}"
        try:
            admin_client.create_queue(queue_name)
            print("Queue created:", queue_name)

            with ServiceBusClient.from_connection_string(conn_str=CONNECTION_STR, logging_enable=True) as servicebus_client:
                with servicebus_client.get_queue_sender(queue_name=queue_name) as sender:
                    send_message(sender)

                with servicebus_client.get_queue_receiver(queue_name=queue_name) as receiver:
                    receive_messages(receiver)
        finally:
            try:
                admin_client.delete_queue(queue_name)
                print("Queue deleted:", queue_name)
            except Exception as e:
                print(f"Error deleting queue ({type(e).__name__}):", e)


if __name__ == "__main__":
    main()

File: Sample-Code-Snippets/JS/azure-service-bus-emulator-node-sample/sample-send-receive.js
import { ServiceBusClient, ServiceBusAdministrationClient } from "@azure/service-bus";
import { randomUUID } from "crypto";

// Configuration
let EMULATOR_HOST = "localhost";
let EMULATOR_HTTP_PORT = 5300;
let connectionString = `Endpoint=sb://${EMULATOR_HOST};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;`;
let adminApiConnectionString = `Endpoint=sb://${EMULATOR_HOST}:${EMULATOR_HTTP_PORT};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;`;
const queueName = `queue-${randomUUID()}`;

function updateConnectionStringsFromEnv() {
    EMULATOR_HOST = process.env.EMULATOR_HOST || "localhost";
    EMULATOR_HTTP_PORT = Number(process.env.EMULATOR_HTTP_PORT || "5300");
    connectionString = `Endpoint=sb://${EMULATOR_HOST};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;`;
    adminApiConnectionString = `Endpoint=sb://${EMULATOR_HOST}:${EMULATOR_HTTP_PORT};SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=SAS_KEY_VALUE;UseDevelopmentEmulator=true;`;
    if (EMULATOR_HTTP_PORT === 80) {
        adminApiConnectionString = connectionString;
    }
    console.log(`Runtime ConnectionString: ${connectionString}`);
    console.log(`Admin API ConnectionString: ${adminApiConnectionString}`);
}

async function waitForServiceBusEmulatorToBeHealthy() {
    const healthUrl = `http://${EMULATOR_HOST}:${EMULATOR_HTTP_PORT}/health`;
    console.log("Waiting for Service Bus Emulator to be healthy...");
    while (true) {
        try {
            const response = await fetch(healthUrl, { method: "GET" });
            if (response.ok) {
                console.log("Service Bus Emulator is healthy.");
                return;
            }
        } catch (err) {
            console.log(`Service Bus Emulator is not healthy yet. health_url=${healthUrl} Continuing to wait...${err}`);
        }
        await new Promise((resolve) => setTimeout(resolve, 1000));
    }
}

async function main() {
    console.log(`--- Starting Run: ${queueName} ---`);

    updateConnectionStringsFromEnv();
    await waitForServiceBusEmulatorToBeHealthy();

    // 1. Initialize the Administration Client
    const adminClient = new ServiceBusAdministrationClient(adminApiConnectionString);

    // Temporary hacks until SDK supports non-TLS and custom-port from connection-string for management operations
    const internalClient = adminClient["_httpClient"]; 
    const originalSendRequest = internalClient.sendRequest.bind(internalClient);
    internalClient.sendRequest = async function (request) {
        request.allowInsecureConnection = true; // Bypass TLS for emulator
        return originalSendRequest(request);
    };
    adminClient.getUrl = function (path, queryParams) {
        const baseUri = `http://${EMULATOR_HOST}:${EMULATOR_HTTP_PORT}/${path}`; // Override to custom admin-api port
        const requestUrl = new URL(baseUri);
        requestUrl.searchParams.set("api-version", this.serviceVersion || "2021-05");
        if (queryParams) {
            for (const [key, value] of Object.entries(queryParams)) {
                requestUrl.searchParams.set(key, value);
            }
        }
        return requestUrl.toString();
    };

    try {
        // Step 1: Create Queue
        console.log(`[Admin] Creating queue: ${queueName}...`);
        await adminClient.createQueue(queueName);
        console.log(`[Admin] Success: Queue created.`);

        // Step 2: Send/Receive (AMQP)
        const sbClient = new ServiceBusClient(connectionString);
        const sender = sbClient.createSender(queueName);
        const receiver = sbClient.createReceiver(queueName);

        console.log(`[Sender] Sending test message...`);
        await sender.sendMessages({ body: "Hello World!" });

        console.log(`[Receiver] Receiving...`);
        const messages = await receiver.receiveMessages(1, { maxWaitTimeInMs: 3000 });
        for (const msg of messages) {
            console.log(`[Receiver] Data:`, msg.body);
            await receiver.completeMessage(msg);
        }

        await sender.close();
        await receiver.close();
        await sbClient.close();

        // Step 3: Delete Queue
        console.log(`[Admin] Deleting queue: ${queueName}...`);
        await adminClient.deleteQueue(queueName);
        console.log(`[Admin] Cleanup complete.`);

    } catch (err) {
        console.error(`\n[Run Failed]`);
        console.error(`Message: ${err.message}`);
    }
}

main();